### PR TITLE
Use a smaller scroll page size to avoid timeouts

### PR DESCRIPTION
**Pull request by esoergel**
https://manage.dimagi.com/default.asp?267392
https://sentry.io/dimagi/commcarehq/issues/428226096/
Looks like a repeat of https://manage.dimagi.com/default.asp?248384#BugEvent.1297392
@emord @dannyroberts pinging you because you worked on this last time

@millerdev @mkangia code buddies

### CODE REVIEW

File: corehq/apps/export/export.py [hunk 1]
```diff
@@ -336,7 +336,8 @@ def get_export_documents(export_instance, filters):
     # We believe we can occasionally hit the 5m limit to process a single scroll window
     # with a window size of 1000 (https://manage.dimagi.com/default.asp?248384).
     # Thus, smaller window size is intentional
-    return query.size(500).scroll()
+    # Another option we have is to bump the "scroll" parameter up from "5m"
+    return query.size(200).scroll()
```
Comment (esoergel) [inline corehq/apps/export/export.py]: As I mentioned in this comment, we can also bump the cache timeout here: https://github.com/dimagi/commcare-hq/blob/master/corehq/elastic.py#L256-L256
Not sure which is preferable.

Comment (emord) [inline corehq/apps/export/export.py]: Comment says that this is 1000

**Review by emord — COMMENTED** [review]
exports is a mess. You own exports now right? I have some ideas to make this better if you want to offline with it. My ideas aren't necessarily trivial though :/

File: corehq/apps/export/export.py (hunk 1 above)
Comment (esoergel) [reply corehq/apps/export/export.py]: I think the comment says that when it was 1000 we had problems, which is why it's now smaller.

**Comment by esoergel** [comment]
@emord 
Ostensibly, yeah.  I haven't actually done any substantial work in it yet though, so not sure how thorough my understanding is.  Whatta you got?

**Comment by emord** [comment]
They're kind of all over the place (this fix seems ok in the short term, but we're going to need to do something soon if we've had to go from 1000 to 200 recently)

1. Rewrite exports :P 
2. Upgrading ES to a more current version should make this better. 
3. Reindex forms and cases with more shards (should probably done after #2) (shards would be more important for ICDS than prod)
4. Structure our forms as "time based data" where you have a different index for a specific timeframe (like one per day/week/month) and use the forms alias to reference all ES indexes. (relevant docs: https://www.elastic.co/guide/en/elasticsearch/guide/current/time-based.html) I'm not sure how cases would need to be structured. (by-domain would work for prod, but ICDS would still be SOL)
5. Change export code to use the commcare-export tool (like instead of directly having a scroll start hitting our supported API that commcare-export uses) and direct more people to using that kind of tool for exports that are large
6. Instead of using scroll, break up the exports into smaller time frames (or user groups or whatever relavant filter we need). 
7. Having a max time frame that you can export from (these exports generally work if you break them into smaller time frames as noted early in the ticket). If you exceed that time frame then you have to use commcare-export or you can use the UI to download the exports over different frames manually

With options 5 & 6 we could also store the current state of the export in postgres and resume the task if the export fails to complete. We could also make it more divide-and-conquer by creating a lot of different export tasks that each export a small time frame with a final task merging all of their results.

@snopoke may have some other ideas. I think he's been doing something custom so that ICDS's exports work

**Comment by esoergel** [comment]
@emord Does this mirror your understanding of the scroll query?

1. We come up with a massive ES query containing all the results we want to export
2. We tell elasticsearch that we're going to be pulling from that query in chunks
3. ES precomputes the whole query and stores it as a temporary search context
4. We pull one page of results from ES
5. We do some processing and transformations in memory, and I assume put it in an open export file
6. We repeat 4 and 5 until the results are exhausted
7. ES drops the search context
8. We serve the export file

My read on this is that step 5 is taking too long, so we sometimes hit the 5 minute timeout.  From the [docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-scroll.html#scroll-search-context):
>  [The timeout] does not need to be long enough to process all data — it just needs to be long enough to process the previous batch of results. Each scroll request (with the scroll parameter) sets a new expiry time.

Given that, I don't think this is a problem with ES, really, since the problem is that the non-ES code takes too long, and ES just gets bored of waiting.  We can address this by reducing the size of each page or by extending the timeout.

Now the whole reason this is an issue at all is because elasticsearch freezes the documents in their current state so it can "return the results of the initial search request, regardless of subsequent changes to documents".  That sounds nice, but I'm not sure how much it matters for this application.  One alternate approach we could do is pull all the doc IDs up front, then pull the docs in chunks.  Then ES wouldn't need to maintain the search context on their end.  The downside of course is that we'd pick up some changes that happen mid export generation.

**Comment by emord** [comment]
FWIW I haven't actually thought about this in a couple of months and never dug too deeply into it. so my understanding may be off or I may be misremembering some of it. My ideas were in a numbered list, but I think all of them could be done independently of one another.

> Now the whole reason this is an issue at all is because elasticsearch freezes the documents in their current state so it can "return the results of the initial search request, regardless of subsequent changes to documents". That sounds nice, but I'm not sure how much it matters for this application

Yeah. I don't think that it matters for exports. I don't think we need these frozen

> ES precomputes the whole query and stores it as a temporary search context

I don't think that it precomputes the whole query. Generally when this happens it fails halfway through the export correct? If it was precomputing it, I would expect it to always fail at the before any docs were pulled

> Given that, I don't think this is a problem with ES, really, since the problem is that the non-ES code takes too long, and ES just gets bored of waiting. We can address this by reducing the size of each page or by extending the timeout.

I think that there are two sides to the issue. 

ES side: As stated above I doubt that ES precomputes the entire query. I think it stores some state so that it can continue scrolling when the next request comes in. At some point in large queries that state gets large/inefficient and subsequent queries will take a long time. This would probably be faster with a better ES data model/ES upgrades. 

We'll be upgrading ES anyways at some point. part of this was probably just optimism that that will help some of our issues.

App side: Agreed that part of this could be the export side taking too long, but I wouldn't expect that the code would take longer on the 100th iteration than the 1st since it should only be concerned with the current docs.

**Comment by snopoke** [comment]
I've started putting this together into a doc: https://docs.google.com/document/d/1Fzn2ph4GTISpOIlfsx-9yn9EMDvdl-UthFS4fpgHk3o/edit?ts=5a70a842#

**Pull request merged**